packages/core/register/src/hook.js

```diff
@@ -0,0 +1,137 @@
+import Module from 'module';
+import process from 'process';
+import path from 'path';
+import fs from 'fs';
+import {addHook} from 'pirates';
+import resolveFrom from 'resolve-from';
+
+import Parcel, {Asset, Dependency, Environment} from '@parcel/core';
+import syncPromise from '@parcel/utils/lib/syncPromise';
+import {loadConfig} from '@parcel/utils/lib/config';
+
+const originalRequire = Module.prototype.require;
+const DEFAULT_CLI_OPTS = {
+  watch: false
+};
+
+let hooks = {};
+let parcelDeps = [];
+
+function isParcelDep(filename) {
+  if (parcelDeps.includes(filename)) {
+    return true;
+  }
+
+  let loadedConfig = syncPromise(loadConfig(filename, ['package.json']));
+  if (loadedConfig === null) {
+    return false;
+  }
+  let pkg = loadedConfig.config;
+  return pkg && (pkg.name.includes('@parcel') || pkg.name.includes('parcel-'));
+}
+
+export default function register(opts = DEFAULT_CLI_OPTS) {
+  // Replace old hook, as this one likely contains options.
+  if (hooks) {
+    for (let extension in hooks) {
+      hooks[extension]();
+    }
+  }
+
+  let parcel = new Parcel({
+    entries: [path.join(process.cwd(), 'index.js')],
+    cliOpts: opts
+  });
+
+  let environment = new Environment({
+    context: 'node',
+    engines: {
+      node: process.versions.node
+    }
+  });
+
+  syncPromise(parcel.init());
+
+  // As Parcel is pretty much fully asynchronous, create an async function and wrap it in a syncPromise later...
+  async function fileProcessor(code, filename) {
+    if (isParcelDep(filename)) {
+      return code;
+    }
+
+    try {
+      let result = await parcel.runTransform({
+        filePath: filename,
+        env: environment
+      });
+
+      if (result.assets && result.assets.length >= 1) {
+        let output = '';
+        let asset = result.assets.find(a => a.type === 'js');
+        if (asset) {
+          asset = new Asset(asset);
```

Comment:
Should already be an asset instance here because of the serializer so no need to construct one

Reply:
Awesome, will update :)